### ISSUE TITLE
GridView doesn't make the cell fully visible when swiping right 

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#36307]
Internal: b/137624348
Related: #61624 for iOS issues.

When using a GridView, and turning TalkBack on, swiping right doesn't make the full grid cell visible if it is already partly visible. Scrolling beyond the list *does* make future rows visible.

Screenshot attached.

Full repro code appended.

Steps: 
Turn on TalkBack, then launch the app. Swipe right, and you'll see that the lowest row of cells isn't brought fully into view when they're highlighted.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  State createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool useSemanticsDebugger = false;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        showSemanticsDebugger: useSemanticsDebugger,
        theme: ThemeData(primaryColor: Colors.blue),
        home: Scaffold(
          appBar: AppBar(
            title: Text('GridView Semantics Test'),
          ),
          body: GridView.count(crossAxisCount: 4,
              children: _buildChildren(context)
          ),
        ));
  }

  List<Widget> _buildChildren(BuildContext context) {
    final widgets = List<Widget>();
    for(int i = 0; i < 100; ++i) {
      widgets.add(ListTile(
        title: Text('Cell $i'),
      ));
    }
    return widgets;
  }
}
```

![image](https://user-images.githubusercontent.com/6687929/61330248-0e84cb80-a7d4-11e9-86ff-1fb3fc4ecd20.png)

**Comment by QuattroNinos** [comment]
Internally tracked at b/140825925

**Comment by nscobie** [comment]
The problems related to this issue are somewhat different on Android vs iOS, so I've split off the iOS problems to issue #61624.

### iOS

As documented on #61624, this problem stems from how implicit accessibility scrolling had to be implemented on iOS, and applies to all scrollable containers, including things like `ListView`.

### Android

Android seems to have better accessibility support for scrolling, but I haven't dug into how we handle it yet. However, it is interesting to note that this problem does not apply to `ListView`s on Android. Additionally, although swiping to the _first_ cell of a partially obscured row in a `GridView` does _not_ scroll the view to make the row visible, swiping further onto the _last_ cell of the row _will_ actually scroll the view to make the cell visible.

**Comment by nscobie** [comment]
Derailed by higher priority issues.

According to this [TODO in AccessibilityBridge.java](https://github.com/flutter/engine/blob/a9910e409c02c214f71f2d6a15fcaf4b4e0b62fc/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L704), it looks like we're currently telling Android that our GridViews are semantically linear lists rather than 2D grids, since the framework currently treats scrollable containers as linear and does not add any metadata about rows vs. columns to the semantics tree. I dug a little into how [GridViews](https://api.flutter.dev/flutter/widgets/GridView-class.html) layout their children with [SliverGrids](https://api.flutter.dev/flutter/widgets/SliverGrid-class.html), which delegate layout to a [SliverGridDelegate](https://api.flutter.dev/flutter/rendering/SliverGridDelegate-class.html).

At the core of this system it looks like the concept of rows and columns are not _strictly_ enforced:

> The tiles can be placed arbitrarily, but it is more efficient to place tiles in roughly in order by scroll offset because grids reify a contiguous sequence of children.

So I'm unsure how we would go about calculating how many rows vs. columns a `SliverGrid` has in a generalizable way.

---

However, even if we could, I'm not sure if that would fix the problem. I tried hardcoding how many rows and columns the AccessibilityBridge told Android my grid contained, and it did not fix the original issue.

I briefly skimmed how [TalkBack is deciding to call showOnScreen](https://github.com/google/talkback/blob/9b1d5132b074174d02886faccf731e814d69363c/talkback/src/main/java/focusmanagement/FocusProcessorForLogicalNavigation.java#L924), and my cursory understanding is it only does so if the very next element would be off screen, which lines up with my observations in the above comment ("although swiping to the _first_ cell of a partially obscured row in a `GridView` does _not_ scroll the view to make the row visible, swiping further onto the _last_ cell of the row _will_ actually scroll the view to make the cell visible."), but unfortunately won't fix the original issue if true. I have not yet tested TalkBack on a GridView in a native app.

---

**As I noted on the [related iOS issue](https://github.com/flutter/flutter/issues/61624#issuecomment-665401888), I am considering a potential cross-platform solution, which would make the framework add a tag to nodes which are partially hidden, similarly to how it currently adds the `isHidden` tag to fully hidden nodes.**

> Now I'm thinking it might be better to add a new isPartiallyHidden flag on the framework, and just do a quick flag check on the engine side. This would help solve similar problems on other platforms and promote code reuse. The current code which calculates isHidden could be extended to check if the paintRect intersection shrunk the semantics rect, so it would just be O(1) time per node on each tree update on the framework side rather than O(K) (where K = depth) whenever a node is focused.
> 
> I may need to write a design doc. Will post more info soon.

**Comment by nscobie** [comment]
> I briefly skimmed how TalkBack is deciding to call showOnScreen, and my cursory understanding is it only does so if the very next element would be off screen, which lines up with my observations in the above comment ("although swiping to the first cell of a partially obscured row in a GridView does not scroll the view to make the row visible, swiping further onto the last cell of the row will actually scroll the view to make the cell visible."), but unfortunately won't fix the original issue if true. I have not yet tested TalkBack on a GridView in a native app.

The grid layout mode for Google Drive, Dropbox, and the default Files app on my Pixel 3a all exhibit the same behavior seen in Flutter, which leads me to believe this is an issue with TalkBack. However, I strongly believe we should fix this issue on the Flutter side, since the same issue is [exhibited on iOS](https://github.com/flutter/flutter/issues/61624) (except with ListViews as well), and likely other platforms (untested).

**Comment by goderbauer** [comment]
> The grid layout mode for Google Drive, Dropbox, and the default Files app on my Pixel 3a all exhibit the same behavior seen in Flutter, which leads me to believe this is an issue with TalkBack.

I was just gonna say, I think what Flutter does is just the native behavior in TalkBack. TalkBack itself doesn't have support for scrolling a list if focus moves to a partially visible element that is not the last/first partially visible element in the list. If this is not desired, it would need to be fixed in TalkBack.

**Comment by nscobie** [comment]
> If this is not desired, it would need to be fixed in TalkBack.

My initial thinking was we could at least temporarily fix this on the Flutter side to help affected customers now, rather than waiting on TalkBack to provide this functionality. However, I definitely see where you're coming from that this should be fixed in TalkBack, and fixing it in Flutter would be a temporary hack (at least for the Android part of things).

The only semi-related TalkBack tracking issue I could find for this is https://b.corp.google.com/issues/126775561 (public), which doesn't quite seem like the same problem. Would you mind doing a brief search internally to see if there's anymore discussion on this which I can't access, and file an issue with TalkBack if not? My summer is coming to a close.

That being said, if they do fix it then I believe this part of my above comment applies, and we'll need to add support for figuring out how many rows vs. columns a GridView has:

> According to this [TODO in AccessibilityBridge.java](https://github.com/flutter/engine/blob/a9910e409c02c214f71f2d6a15fcaf4b4e0b62fc/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L704), it looks like we're currently telling Android that our GridViews are semantically linear lists rather than 2D grids, since the framework currently treats scrollable containers as linear and does not add any metadata about rows vs. columns to the semantics tree. I dug a little into how [GridViews](https://api.flutter.dev/flutter/widgets/GridView-class.html) layout their children with [SliverGrids](https://api.flutter.dev/flutter/widgets/SliverGrid-class.html), which delegate layout to a [SliverGridDelegate](https://api.flutter.dev/flutter/rendering/SliverGridDelegate-class.html).

> At the core of this system it looks like the concept of rows and columns are not _strictly_ enforced:

> > The tiles can be placed arbitrarily, but it is more efficient to place tiles in roughly in order by scroll offset because grids reify a contiguous sequence of children.

> So I'm unsure how we would go about calculating how many rows vs. columns a `SliverGrid` has in a generalizable way.

As for this issue on our tracker, should it be closed as WAI/"can't fix", or potentially left open to track a TalkBack issue?

**Comment by goderbauer** [comment]
> Would you mind doing a brief search internally to see if there's anymore discussion on this which I can't access, and file an issue with TalkBack if not?

I'll check.

> As for this issue on our tracker, should it be closed as WAI/"can't fix", or potentially left open to track a TalkBack issue?

I would close this as WAI with a note that its on talkback.
